sdk/helper/clientcountutil/clientcountutil.go

```diff
@@ -0,0 +1,385 @@
+	s := &generation.Segment{
+		Clients: &generation.Clients{},
+	}
+	for _, opt := range opts {
+		opt(s)
+	}
+	if d.addingToMonth.GetSegments() == nil {
+		d.addingToMonth.Clients = &generation.Data_Segments{Segments: &generation.Segments{}}
+	}
+	d.addingToMonth.GetSegments().Segments = append(d.addingToMonth.GetSegments().Segments, s)
+	d.addingToSegment = s
+	return d
+}
+
+// ToJSON returns the JSON representation of the data
+func (d *ActivityLogDataGenerator) ToJSON() ([]byte, error) {
+	return protojson.Marshal(d.data)
+}
+
+// ToProto returns the ActivityLogMockInput protobuf
+func (d *ActivityLogDataGenerator) ToProto() *generation.ActivityLogMockInput {
+	return d.data
+}
+
+// Write writes the data to the API with the given write options. The method
+// returns the new paths that have been written. Note that the API endpoint will
+// only be present when Vault has been compiled with the "testonly" flag.
+func (d *ActivityLogDataGenerator) Write(ctx context.Context, writeOptions ...generation.WriteOptions) ([]string, error) {
+	d.data.Write = writeOptions
+	err := VerifyInput(d.data)
+	if err != nil {
+		return nil, err
+	}
+	data, err := d.ToJSON()
+	if err != nil {
+		return nil, err
+	}
+	resp, err := d.client.Logical().WriteBytesWithContext(ctx, "sys/internal/counters/activity/write", data)
+	if err != nil {
+		return nil, err
+	}
+	paths := resp.Data["paths"]
+	castedPaths, ok := paths.([]interface{})
+	if !ok {
+		return nil, fmt.Errorf("invalid paths data: %v", paths)
+	}
+	returnPaths := make([]string, 0, len(castedPaths))
+	for _, path := range castedPaths {
+		returnPaths = append(returnPaths, path.(string))
+	}
+	return returnPaths, nil
+}
+
+// VerifyInput checks that the input data is valid
+func VerifyInput(input *generation.ActivityLogMockInput) error {
+	// mapping from monthsAgo to the month's data
+	months := make(map[int32]*generation.Data)
+
+	// this keeps track of the index of the earliest month. We need to verify
+	// that this month doesn't have any repeated clients
+	earliestMonthsAgo := int32(0)
+
+	// this map holds a set of the month indexes for any RepeatedFromMonth
+	// values. Each element will be checked to ensure month that should be
+	// repeated from exists in the input data
+	repeatedFromMonths := make(map[int32]struct{})
+
+	for _, month := range input.Data {
+		monthsAgo := month.GetMonthsAgo()
+		if monthsAgo > earliestMonthsAgo {
+			earliestMonthsAgo = monthsAgo
+		}
+
+		// verify that no monthsAgo value is repeated
+		if _, seen := months[monthsAgo]; seen {
+			return fmt.Errorf("multiple months with monthsAgo %d", monthsAgo)
+		}
+		months[monthsAgo] = month
+
+		// the number of segments should be correct
+		if month.NumSegments > 0 && int(month.NumSegments)-len(month.GetSkipSegmentIndexes())-len(month.GetEmptySegmentIndexes()) <= 0 {
+			return fmt.Errorf("number of segments %d is too small. It must be large enough to include the empty and skipped segments", month.NumSegments)
```

Comment:
nit: would it be useful to include the number of empty and skipped segments in the error for reference?

Reply:
added!